src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,465 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+};
+
+use common_daft_config::DaftExecutionConfig;
+use common_error::DaftResult;
+use common_file_formats::{FileFormatConfig, ParquetSourceConfig};
+use common_runtime::get_io_runtime;
+use daft_io::{get_io_client, IOStatsContext};
+use daft_parquet::read::read_parquet_metadata;
+use daft_stats::TableMetadata;
+use futures::future::try_join_all;
+use itertools::Itertools;
+use parquet2::metadata::{FileMetaData, RowGroupList};
+
+use super::BoxScanTaskIter;
+use crate::{storage_config::StorageConfig, ChunkSpec, DataSource, ScanTask, ScanTaskRef};
+
+enum SplitSingleParquetFileByRowGroupsState {
+    Accumulating(usize, Vec<usize>, f64),
+    ReadyToEmit(usize, Vec<usize>),
+}
+
+/// Splits a single Parquet-based ScanTask by rowgroups into smaller chunked ScanTasks
+///
+/// This is used as an Iterator, yielding ScanTaskRefs after splitting the Parquet file
+struct ParquetFileSplitter<'a> {
+    state: SplitSingleParquetFileByRowGroupsState,
+    scan_task: &'a ScanTask,
+    file_metadata: FileMetaData,
+    config: &'a DaftExecutionConfig,
+
+    // Call once and cache this, since it can be expensive when called in a loop
+    scan_task_in_memory_size_estimate: Option<usize>,
+}
+
+impl<'a> ParquetFileSplitter<'a> {
+    pub fn new(
+        scan_task: &'a ScanTask,
+        file_metadata: FileMetaData,
+        config: &'a DaftExecutionConfig,
+    ) -> Self {
+        Self {
+            state: SplitSingleParquetFileByRowGroupsState::Accumulating(0, Vec::new(), 0.),
+            scan_task,
+            file_metadata,
+            config,
+            scan_task_in_memory_size_estimate: scan_task
+                .estimate_in_memory_size_bytes(Some(config)),
+        }
+    }
+}
+
+impl<'a> Iterator for ParquetFileSplitter<'a> {
+    type Item = ScanTaskRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match &mut self.state {
+                SplitSingleParquetFileByRowGroupsState::Accumulating(
+                    idx,
+                    accumulated,
+                    estimated_accumulated_size_bytes,
+                ) => {
+                    let idx = *idx;
+
+                    // Finished iterating: idx exceeds the number of available rowgroups
+                    if idx >= self.file_metadata.row_groups.len() {
+                        if accumulated.is_empty() {
+                            return None;
+                        } else {
+                            self.state = SplitSingleParquetFileByRowGroupsState::ReadyToEmit(
```

Comment:
these transitions can actually be methods on the state itself.

```
if self.state.current_index() >= self.file_metadata.row_groups.len() {
  self.state.mark_ready_to_emit()
}


```

Reply:
also you can likely pass the `total row groups` in the state, so functionality can be encompassed completely in the state without have to do `self.state.current_index() >= self.file_metadata.row_groups.len()`

Reply:
instead you could do 

`while !self.state.is_depleted() && !self.state.is_ready() {`

for the main loop.

Overall i would advise that you expose an "interface" to your consumer `ParquetFileSplitter` rather than digging into the internals.

Reply:
Made changes to remove the state machine abstraction in favor of a simpler accumulator. Should be much simpler now.